cpp/src/io/json/json_tree.cu

```diff
@@ -487,13 +487,54 @@ std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_n
   device_span<TreeDepthT const> node_levels,
   device_span<size_type const> node_type,
   device_span<NodeIndexT const> parent_node_ids,
+  bool is_array_of_arrays,
+  bool is_enabled_lines,
   rmm::cuda_stream_view stream,
   rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
   auto const num_nodes = parent_node_ids.size();
   rmm::device_uvector<size_type> col_id(num_nodes, stream, mr);
 
+  // if level 1 is a list, then it should take its list index as key!
+  // (level 1 for values, level 0 for values-JSONLines)
+  // copy nodes of level 1 children. (level 2) and their parent_node_id
+  // exclusive scan by key -> get their indices and scatter to their node id
+  // use this index for hashing at level 2
+
+  // array of arrays
+  NodeIndexT const row_array_children_level = is_enabled_lines ? 1 : 2;
+  rmm::device_uvector<size_type> list_indices(0, stream);
+  if (is_array_of_arrays) {
+    auto num_level2_nodes = thrust::count(
+      rmm::exec_policy(stream), node_levels.begin(), node_levels.end(), row_array_children_level);
+    rmm::device_uvector<NodeIndexT> level2_nodes(num_level2_nodes, stream);
+    rmm::device_uvector<NodeIndexT> level2_indices(num_level2_nodes, stream);
+    auto const iter = thrust::copy_if(rmm::exec_policy(stream),
+                                      thrust::counting_iterator<NodeIndexT>(0),
+                                      thrust::counting_iterator<NodeIndexT>(num_nodes),
+                                      node_levels.begin(),
+                                      level2_nodes.begin(),
+                                      [row_array_children_level] __device__(auto level) {
+                                        return level == row_array_children_level;
+                                      });
+    // memory usage could be reduced by using different data structure (hashmap)
+    // or alternate method to hash it at node_type
+    auto level2_parent_nodes =
+      thrust::make_permutation_iterator(parent_node_ids.begin(), level2_nodes.cbegin());
+    thrust::exclusive_scan_by_key(rmm::exec_policy(stream),
+                                  level2_parent_nodes,
+                                  level2_parent_nodes + num_level2_nodes,
+                                  thrust::make_constant_iterator(NodeIndexT{1}),
+                                  level2_indices.begin());
+    list_indices.resize(num_nodes, stream);
+    thrust::scatter(rmm::exec_policy(stream),
+                    level2_indices.cbegin(),
+                    level2_indices.cend(),
+                    level2_nodes.cbegin(),
+                    list_indices.begin());
+  }
```

Comment:
Could we just call `get_values_column_indices` instead? There seems to be a lot of redundancy between the two.

Reply:
`get_values_column_indices` takes `col_ids` (column ids) as input. At`records_orient_tree_traversal`, col_ids are not generated yet. So, we can not call `get_values_column_indices`. As suggested, redundancy can be reduced.

Reply:
Understood. I was implying to further parametrize `get_values_column_indices` such that it can serve both use cases. But I think the way you have it now, through extracting `get_array_children_indices`, is even better, thanks 🙏